File: benchmarks/c/singlethread.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <unistd.h>

int main(int argc, char **argv) {
  if (argc < 3) {
    fprintf(stderr, "Usage: %s <filepath> <size>\n", argv[0]);
    return 1;
  }

  int size = atoi(argv[2]);
  if (size <= 0) {
    fprintf(stderr, "Error: Size must be a positive integer\n");
    return 1;
  }
  int half_size = size / 2;

  int fd = open(argv[1], O_RDONLY);
  if (fd == -1) {
    fprintf(stderr, "Error: Could not open file %s\n", argv[1]);
    return 1;
  }

  int *data = mmap(NULL, size * sizeof(int), PROT_READ, MAP_PRIVATE, fd, 0);
  if (data == MAP_FAILED) {
    fprintf(stderr, "Error: Memory mapping failed\n");
    close(fd);
    return 1;
  }

  double result = 0;
  for (size_t i = 0; i < half_size; i++) {
    result += sqrt(fabs(cos(data[i]) * sin(data[half_size + i])));
  }

  printf("%f\n", result);

  munmap(data, size * sizeof(int));
  close(fd);

  return 0;
}

File: benchmarks/java/SingleThread.java
import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;

public class SingleThread {
    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("Usage: java SingleThread <filepath> <size>");
            System.exit(1);
        }

        try {
            int size = Integer.parseInt(args[1]);
            if (size <= 0) {
                System.err.println("Error: Size must be a positive integer");
                System.exit(1);
            }
            int halfSize = size / 2;

            File file = new File(args[0]);
            RandomAccessFile raf = new RandomAccessFile(file, "r");
            FileChannel channel = raf.getChannel();

            // Map the file into memory
            MappedByteBuffer buffer = channel.map(
                FileChannel.MapMode.READ_ONLY,
                0,
                size * Integer.BYTES
            );

            double result = 0;
            for (int i = 0; i < halfSize; i++) {
                buffer.order(java.nio.ByteOrder.LITTLE_ENDIAN);
                int firstValue = buffer.getInt(i * Integer.BYTES);
                int secondValue = buffer.getInt((halfSize + i) * Integer.BYTES);
                result += Math.sqrt(Math.abs(Math.cos(firstValue) * Math.sin(secondValue)));
            }

            System.out.printf("%f\n", result);

            // Clean up
            channel.close();
            raf.close();

        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        } catch (NumberFormatException e) {
            System.err.println("Error: Size must be a valid integer");
            System.exit(1);
        }
    }
}
